File: golang/1736_maximum_time/maximum_time.go
/*
 * @Date: 2021-07-24 13:31:56
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-07-24 13:38:44
 */

package main

func maximumTime(time string) string {
	t := []byte(time)
	if t[0] == '?' {
		if '4' <= t[1] && t[1] <= '9' {
			t[0] = '1'
		} else {
			t[0] = '2'
		}
	}
	if t[1] == '?' {
		if t[0] == '2' {
			t[1] = '3'
		} else {
			t[1] = '9'
		}
	}
	if t[3] == '?' {
		t[3] = '5'
	}
	if t[4] == '?' {
		t[4] = '9'
	}
	return string(t)
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed")
		}
	}
	{
		time := "2?:?0"
		assert(maximumTime(time) == "23:50")
	}

	{
		time := "0?:3?"
		assert(maximumTime(time) == "09:39")
	}

	{
		time := "1?:22"
		assert(maximumTime(time) == "19:22")
	}
}

File: cpp/1736_maximum_time/maximum_time.cpp
/*
 * @Date: 2021-07-24 13:31:56
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-07-24 13:38:44
 */

#include <cassert>
#include <string>

using namespace std;

class Solution {
public:
  string maximumTime(string time) {
    if (time[0] == '?') {
      time[0] = ('4' <= time[1] && time[1] <= '9') ? '1' : '2';
    }
    if (time[1] == '?') {
      time[1] = (time[0] == '2') ? '3' : '9';
    }
    if (time[3] == '?') {
      time[3] = '5';
    }
    if (time[4] == '?') {
      time[4] = '9';
    }
    return time;
  }
};

int main() {
  {
    string time = "2?:?0";
    assert(Solution{}.maximumTime(time) == "23:50");
  }

  {
    string time = "0?:3?";
    assert(Solution{}.maximumTime(time) == "09:39");
  }

  {
    string time = "1?:22";
    assert(Solution{}.maximumTime(time) == "19:22");
  }

  return 0;
}